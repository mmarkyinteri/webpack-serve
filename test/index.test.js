const { readFileSync: read } = require('fs');
const path = require('path');

const request = require('supertest');

const serve = require('../lib/index');

jest.setTimeout(10000);

describe('serve', () => {
  test('serve', () => {
    const argv = { logLevel: 'silent' };
    const opts = {};
    return serve(argv, opts).then(({ app, options }) => {
      expect(app).toMatchSnapshot();
      expect(options).toMatchSnapshot({
        compiler: expect.any(Object),
      });

      return new Promise((resolve) => {
        setTimeout(() => app.stop(resolve), 500);
      });
    });
  });

  test('bad argv', () => {
    const argv = [];
    const opts = {};
    const fn = () => serve(argv, opts);

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  test('no params', () =>
    serve().then(
      ({ app }) =>
        new Promise((resolve) => {
          app.stop(resolve);
        })
    ));

  test('basic config', () => {
    const argv = { logLevel: 'silent' };
    const opts = { config: require('./fixtures/basic/webpack.config') };
    return serve(argv, opts).then(({ app, on, options }) => {
      const { server } = app;
      expect(app).toMatchSnapshot();
      expect(options).toMatchSnapshot({
        compiler: expect.any(Object),
      });

      return new Promise((resolve) => {
        on('build-finished', ({ stats }) => {
          expect(stats).toBeDefined();
          request(server)
            .get('/output.js')
            .expect(200)
            .then((response) => {
              expect(response.text.length).toBeGreaterThan(1000);
              app.stop(resolve);
            });
        });
      });
    });
  });

  test('config + add', () => {
    const argv = { logLevel: 'silent' };
    const opts = {
      add(app, middleware) {
        middleware.webpack().then(() => {
          middleware.content({
            index: 'index.htm',
          });

          app.use((ctx) => {
            // eslint-disable-next-line no-param-reassign
            ctx.status = 204;
          });
        });
      },
      config: require('./fixtures/htm/webpack.config'),
    };

    return serve(argv, opts).then(({ app }) => {
      const req = request(app.server);
      return Promise.all([
        // credit: @frenzzy for this test
        req.get('/fallthrough').expect(204),
        req.get('/index.htm').expect(200),
        req.get('/output.js').expect(200),
      ]).then(() => new Promise((resolve) => app.stop(resolve)));
    });
  });

  test('https', () => {
    process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

    const argv = { logLevel: 'silent' };
    const cert = read(path.resolve(__dirname, './fixtures/test-cert.pem'));
    const key = read(path.resolve(__dirname, './fixtures/test-key.pem'));
    const opts = {
      config: require('./fixtures/multi/webpack.config'),
      https: { cert, key },
    };
    return serve(argv, opts).then(({ app }) => {
      const { port } = app.server.address();
      const req = request(`https://localhost:${port}`);
      return Promise.all([
        req.get('/static/client.js').expect(200),
        req.get('/server/server.js').expect(200),
      ]).then(() => new Promise((resolve) => app.stop(resolve)));
    });
  });

  test('multi config', () => {
    const argv = { logLevel: 'silent' };
    const opts = { config: require('./fixtures/multi/webpack.config') };
    return serve(argv, opts).then(({ app }) => {
      const req = request(app.server);
      return Promise.all([
        req.get('/static/client.js').expect(200),
        req.get('/server/server.js').expect(200),
      ]).then(() => new Promise((resolve) => app.stop(resolve)));
    });
  });

  test('multi-named config', () => {
    const argv = { logLevel: 'silent' };
    const opts = { config: require('./fixtures/multi-named/webpack.config') };
    return serve(argv, opts).then(({ app }) => {
      const req = request(app.server);
      return Promise.all([
        req.get('/bundle1.js').expect(200),
        req.get('/bundle2.js').expect(200),
      ]).then(() => new Promise((resolve) => app.stop(resolve)));
    });
  });

  test('webpack 4 defaults config', () => {
    const argv = { logLevel: 'silent' };
    const opts = {
      config: require('./fixtures/webpack-4-defaults/webpack.config'),
    };
    return serve(argv, opts).then(({ app }) => {
      const req = request(app.server);
      return Promise.all([
        req.get('/index.html').expect(200),
        req.get('/main.js').expect(200),
      ]).then(() => new Promise((resolve) => app.stop(resolve)));
    });
  });
});
