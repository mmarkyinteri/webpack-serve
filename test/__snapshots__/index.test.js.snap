// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serve bad argv 1`] = `"webpack-serve: The \`argv\` parameter must be an object"`;

exports[`serve basic config 1`] = `
Object {
  "env": "test",
  "proxy": false,
  "subdomainOffset": 2,
}
`;

exports[`serve basic config 2`] = `
Object {
  "bus": Nanobus {
    "_listeners": Object {},
    "_name": "nanobus",
    "_starListeners": Array [],
  },
  "clipboard": true,
  "compiler": Any<Object>,
  "content": Array [
    "<PROJECT_ROOT>/test/fixtures/basic",
  ],
  "devMiddleware": Object {
    "logLevel": "silent",
    "publicPath": "/",
  },
  "host": "localhost",
  "hotClient": Object {
    "host": "localhost",
    "logLevel": "silent",
  },
  "http2": false,
  "https": null,
  "logLevel": "silent",
  "logTime": false,
  "open": false,
  "port": 8080,
  "protocol": "http",
}
`;

exports[`serve serve 1`] = `
Object {
  "env": "test",
  "proxy": false,
  "subdomainOffset": 2,
}
`;

exports[`serve serve 2`] = `
Object {
  "bus": Nanobus {
    "_listeners": Object {},
    "_name": "nanobus",
    "_starListeners": Array [],
  },
  "clipboard": true,
  "compiler": Any<Object>,
  "content": Array [
    "<PROJECT_ROOT>",
  ],
  "devMiddleware": Object {
    "logLevel": "silent",
    "publicPath": "/",
  },
  "host": "localhost",
  "hotClient": Object {
    "host": "localhost",
    "logLevel": "silent",
  },
  "http2": false,
  "https": null,
  "logLevel": "silent",
  "logTime": false,
  "open": false,
  "port": 8080,
  "protocol": "http",
}
`;
